tinygrad/renderer/assembly.py

```diff
@@ -153,7 +154,8 @@ def _cast(a, dtype:DType, atype:DType, bitcast=False, u=None, pred=False):
     for u in uops:
       uop,dtype,src,args = u.op,u.dtype,u.src,u.arg
       if uop is UOps.IF:
-        kk(*self.render_bra(f"IF_{r[src[0]][1:]}_{uops.index(u)}", _cast(r[src[0]], dtypes.bool, src[0].dtype, u=u, pred=True)))
+        pred_reg = _cast(r[src[0]], dtypes.bool, src[0].dtype, u=u, pred=True)
```

Comment:
can invert be a rule in extra_mather?
Also I think that pred=True doesn't do anything.

Reply:
Thanks for the feedback! Quick update -- I'm not yet sure why, but when I try to add a rule to the ptx_matcher to invert the IF predicate with:

```
  # invert predicate for UOps.IF
  (UPat((UOps.IF,), name="x", allow_any_len=True, src=(UPat(dtype=dtypes.bool),)),
   lambda x: UOp(x.op, dtype=x.dtype, src=(x.src[0].logical_not(),) + x.src[1:], arg=x.arg)),
```

the ptx unit tests hang. I'll investigate this further tonight when i can use a gpu. As for the pred=True, I've removed that.

And in terms of the llvm renderer, it previously wasn't handling IF/ENDIFs at all (they are only ever created for the new gated stores, i believe), and I think the stack is necessary to handle nested IF/ENDIFs -- if you do think it's too complex, I can work on that more today and try to simplify it further.